### PR TITLE
feat: Implement tests for metadata system events

### LINKED ISSUE #695: Implement automation test for device system event
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for Core Metadata

### Description [**REQUIRED**]
 
This is a placeholder for implementing test for https://docs.edgexfoundry.org/2.3/design/adr/0024-system-events/
There will be more details once this ADR is implemented.

### PR DESCRIPTION
**Pull request by jumpingliu**
close #695
Signed-off-by: melody <melody@iotechsys.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

**Comment by jumpingliu** [comment]
@cherrycl Please help review when you have time, thanks.

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot [hunk 1]
```diff
@@ -1,33 +1,82 @@
...
+    When Create Device For device-virtual With Name ${device_name}
+    Then Single System Event Should Match With Topic core-metadata and device and add
+    [Teardown]  Delete device by name ${device_name}
+
+SystemEvents002-Add Multiple Devices and Receive Correct System Events
+    Given Set Test Variable  ${add_devices_topic}  edgex/system-events/core-metadata/device/add/device-virtual/Virtual-Sample-Profile
+    And Set Test Variable  ${numbers}  3
+    And Run MQTT Subscriber Progress And Output  ${add_devices_topic}  Payload  3
+    When Create 3 Devices For device-virtual
+    Then Multiple ${numbers} System Events Should Match With Topic core-metadata and device and add
```
Comment (cherrycl) [inline TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot:26]:
```suggestion
    Then Multiple 3 System Events Should Match With Topic core-metadata and device and add
```
Same as SystemEvents004.

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot [hunk 2]
```diff
@@ -1,33 +1,82 @@
...
+    Given Set Test Variable  ${add_device_topic}  edgex/system-events/core-metadata/device/add/device-virtual/Virtual-Sample-Profile
+    And Set Test Variable  ${device_name}  add-system-event
+    And Run MQTT Subscriber Progress And Output  ${add_device_topic}  Payload
+    When Create Device For device-virtual With Name ${device_name}
+    Then Single System Event Should Match With Topic core-metadata and device and add
+    [Teardown]  Delete device by name ${device_name}
+
+SystemEvents002-Add Multiple Devices and Receive Correct System Events
+    Given Set Test Variable  ${add_devices_topic}  edgex/system-events/core-metadata/device/add/device-virtual/Virtual-Sample-Profile
+    And Set Test Variable  ${numbers}  3
```
Comment (cherrycl) [inline TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot:23]: Remove. Same as SystemEvents004.

**Review by cherrycl — COMMENTED** [review]
Add terminal keyword on Teardown of every test. Reference [here](https://github.com/edgexfoundry/edgex-taf/blob/c688dc53284221c6b4f37ea79b2c3d2c2813adab/TAF/testScenarios/integrationTest/UC_metrics/app_services_metrics_redis.robot#L35)

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot [hunk 3]
```diff
@@ -1,33 +1,82 @@
 *** Settings ***
-Force Tags     Skipped
+Resource         TAF/testCaseModules/keywords/core-metadata/coreMetadataAPI.robot
+Resource     TAF/testCaseModules/keywords/common/commonKeywords.robot
+Resource  TAF/testCaseModules/keywords/device-sdk/deviceServiceAPI.robot
+Resource  TAF/testCaseModules/keywords/core-command/coreCommandAPI.robot
+Force Tags     backward-skip MessageQueue=MQTT
```
[the remark this answers is not in the document]
Comment (jumpingliu) [reply TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot:6]: updated

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot (hunk 1 above)
Comment (jumpingliu) [reply TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot:26]: updated

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot (hunk 2 above)
Comment (jumpingliu) [reply TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot:23]: updated

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot [hunk 4]
```diff
@@ -1,33 +1,82 @@
...
+    Update Device Command-Device-0 With operatingState=DOWN
+    Update Device Command-Device-1 With adminState=LOCKED
+    Update Device Command-Device-2 With label=test-system-event
+
+Single System Event Should Match With Topic ${source} and ${type} and ${action}
+    ${received_content}  grep file  ${WORK_DIR}/TAF/testArtifacts/logs/${subscriber_file}  Payload
+    Should Not Be Empty  ${received_content}
+    ${encode_system_events}  Evaluate  json.loads('''${received_content}''')
+    ${decode_system_events}  Evaluate  base64.b64decode('${encode_system_events}[Payload]').decode('utf-8')  modules=base64
+    ${payload}  Evaluate  json.loads('''${decode_system_events}''')
```
[the remark this answers is not in the document]
Comment (jumpingliu) [reply TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events.robot:66]: updated

**Comment by jumpingliu** [comment]
review comment addressed, also rename the filename to `metadata_system_event_mqtt.robot` for the upcoming `metadata_system_event_redis.robot`

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    Then Single System Event Should Match With Topic core-metadata and device and delete
+    [Teardown]   Terminate All Processes  kill=True
+
+*** Keywords ***
+Update Multiple Devices to receive system events
+    Update Device Command-Device-0 With operatingState=DOWN
+    Update Device Command-Device-1 With adminState=LOCKED
+    Update Device Command-Device-2 With label=test-system-event
+
+Single System Event Should Match With Topic ${source} and ${type} and ${action}
```
Comment (cherrycl) [inline TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot:64]:
```suggestion
Single System Event Should Match With Source ${source}, Type ${type} and Action ${action}
```
Same as the next keyword.

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    Given Set Test Variable  ${delete_device_topic}  edgex/system-events/core-metadata/device/delete/device-virtual/Virtual-Sample-Profile
+    And Set Test Variable  ${device_name}  delete-system-event
+    And Run MQTT Subscriber Progress And Output  ${delete_device_topic}  Payload
+    And Create Device For device-virtual With Name ${device_name}
+    When Delete device by name ${device_name}
+    Then Single System Event Should Match With Topic core-metadata and device and delete
+    [Teardown]   Terminate All Processes  kill=True
+
+*** Keywords ***
+Update Multiple Devices to receive system events
```
Comment (cherrycl) [inline TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot:59]:
```suggestion
Update Multiple Devices
```

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot (hunk 5 above)
Comment (jumpingliu) [reply TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot]: modified

File: TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot (hunk 6 above)
Comment (jumpingliu) [reply TAF/testScenarios/integrationTest/UC_system_events/metadata_system_events_mqtt.robot]: modified

**Review by cherrycl — APPROVED** [review]
LGTM

**Pull request merged**